atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/async/CassandraAsyncKeyValueService.java

```diff
@@ -78,12 +85,17 @@ private ListenableFuture<Optional<Value>> getCellAsync(TableReference tableRefer
                 .humanReadableTimestamp(timestamp)
                 .build();
 
-        return cqlClient.executeQuery(new GetQuerySpec(queryContext, getQueryParameters));
+        return cqlClientContainer.get().executeQuery(new GetQuerySpec(queryContext, getQueryParameters));
     }
 
     @Override
     public void close() {
-        cqlClient.close();
+        cqlClientContainer.close();
         futuresCombiner.close();
     }
+
+    @Override
+    public boolean isValid() {
+        return !cqlClientContainer.isClosed() && cqlClientContainer.get().isValid();
```

Comment:
Do we have to worry about this race condition?
- `cqlClientContainer.isClosed()` returns false - the container is not closed
- someone else closes the container
- we _then_ execute `cqlClientContainer.get()`, but this throws a `SafeIllegalStateException`

This should be rare, but in this case I think we should return false.